File: C++/AnagramSorter.cpp
/* Given a list of words, group anagrams together
Input: str array Output: 2D str array
Constraints: 1 ≤ strs.length ≤ 10⁴, 1 ≤ strs[i].length ≤ 100, & strs[i] consists of lowercase letters
 */
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

class AnagramSorter {
public:
    std::vector<std:: vector<string>> solution(std::vector<string>& strs){
        unordered_map<string, std:: vector<string>> map;

        for(int i = 0; i < strs.size(); i++){
            string word = strs[i];
            int count[26] = {0};

            for(int j = 0; j < strs.size(); j++){
                count[word[j] - 'a']++;
            }

            string key;
            for(int j = 0; j < 26; j++){
                key += "#";
                key += to_string(count[j]);
            }
            map[key].push_back(word);
        }

        std:: vector<vector<string>> groupings;
        for(auto& entry: map){
            groupings.push_back(entry.second);
        }
        return groupings;
    }
};

int main() {
    vector<string> input = {"eat", "tea", "tan", "ate", "nat", "bat"};

    AnagramSorter sorter;
    vector<vector<string>> result = sorter.solution(input);

    cout << "Grouped Anagrams:\n";
    for (int i = 0; i < result.size(); i++) {
        for (int j = 0; j < result[i].size(); j++) {
            cout << result[i][j] << " ";
        }
        cout << endl;
    }

    return 0;
}

File: Java/AnagramSorter.java
/* Given a list of words, group anagrams together
Input: str array Output: 2D str array
Constraints: 1 ≤ strs.length ≤ 10⁴, 1 ≤ strs[i].length ≤ 100, & strs[i] consists of lowercase letters
 */
import java.util.*;

class AnagramSorter {
    public List<List<String>> solution(String[] strs){
        Map<String, List<String>> map = new HashMap<>();

        for (String word : strs) {
            // Intilize needed vars & count letters of current word
            int[] count = new int[26];
            for(int i = 0; i < word.length(); i++){
                count[word.charAt(i) - 'a']++;
            }
            // Make Unique key based on prev counts, then add words to map
            StringBuilder sb = new StringBuilder();
            for(int i = 0; i < 26; i++){
                sb.append('#');
                sb.append(count[i]);
            }
            String key = sb.toString();
            if(!map.containsKey(key)){
                map.put(key, new ArrayList<>());
            }
            map.get(key).add(word);
        }

        // Map vals -> list of lists & return
        List<List<String>> groupings = new ArrayList<>();
        for(Map.Entry<String, List<String>> entry: map.entrySet()){
            groupings.add(entry.getValue());
        }
        return groupings;
    }

    public static void main(String[] args) {
        String[] input = {"eat", "tea", "tan", "ate", "nat", "bat"};

        AnagramSorter sorter = new AnagramSorter();
        List<List<String>> result = sorter.solution(input);

        System.out.println("Grouped Anagrams:");
        for (int i = 0; i < result.size(); i++) {
            System.out.println(result.get(i));
        }
    }
}
